Close window on auth callback error and show error to user

mifi opened the pull request:
Because simply receiving a "Bad request" is very bad UX
For example if the user cancels the auth flow, or doesn't give companion access
it's better to go back to the uppy screen so they can try again
UPDATE: also show message in UI

aduh95 commented on packages/@uppy/companion/src/server/controllers/callback.js:
```diff
@@ -4,6 +4,18 @@
+        window.close()
+      </script>
+  </head>
+  <body></body>
```
IMHO it would be worth displaying something in the body in case the browser refuses to close the window:
```suggestion
  <body>Authentication failed.</body>
```

mifi replied:
Should we i18n this message?

aduh95 replied:
If we have access to the user's preferred language at this point, sure why not. We should keep in mind that unless the browser does something weird, users would never see this message so it may not be worth it.

mifi replied:
Yea I agree. I don't see any other i18n code in companion.

mifi replied:
BTW in the success scenario, maybe we should do the same.

aduh95 commented on packages/@uppy/companion/src/server/controllers/callback.js:
```diff
@@ -27,5 +39,6 @@ module.exports = function callback (req, res, next) { // eslint-disable-line no-
   logger.debug(`Did not receive access token for provider ${providerName}`, null, req.id)
   logger.debug(grant.response, 'callback.oauth.resp', req.id)
-  return res.sendStatus(400)
+
```
Should we keep the status code to `400`? It feels wrong to send `200` when there's clearly an error.
```suggestion
res.statusCode = 400
```

mifi replied:
I'm not sure whether some browsers will show a different page or error document if the code is not 400, do you know?

aduh95 replied:
Browsers render the HTML file whether the error code is 200 or something else (e.g.: https://uppy.io/page_that_does_not_exist), with the exception of 3xx error code that redirect the user before the page is displayed. There might be other exceptions that I don't know, but I don't think 400 is one of them.

mifi replied:
I have some faint memory that Internet explorer shows a custom error page when it receives an error http status code, but I may be wrong, and I'm not sure if we support IE any more

mifi commented on packages/@uppy/companion/src/server/controllers/callback.js:
```diff
@@ -27,5 +39,6 @@ module.exports = function callback (req, res, next) { // eslint-disable-line no-
   logger.debug(`Did not receive access token for provider ${providerName}`, null, req.id)
   logger.debug(grant.response, 'callback.oauth.resp', req.id)
-  return res.sendStatus(400)
+res.statusCode = 400
```
```suggestion
  res.statusCode = 400
```

mifi commented on packages/@uppy/companion/src/server/controllers/callback.js:
```diff
@@ -27,5 +39,6 @@ module.exports = function callback (req, res, next) { // eslint-disable-line no-
   logger.debug(grant.response, 'callback.oauth.resp', req.id)
-  return res.sendStatus(400)
+res.statusCode = 400
+  return res.send(closePageHtml)
```
I think statusCode also works though? from the link you sent: `Sets the HTTP status for the response. It is a chainable alias of Node’s response.statusCode.`

aduh95 replied:
I'm unsure Express exposes Node’s response.statusCode directly or only through the `.status()` method. It's probably safer to use the same syntax as in the docs example, I don't know Express that well so please disregard if you're confident with the previous suggestion.

mifi replied:
I agree, let's go with `res.status(400).send(closePageHtml)`, it looks more semantic

mifi commented:
I have now also added some extra code to show a message to the user when authentication fails / is aborted.
I added a new english string `authAborted`. How do we get this translated?

<img width="434" alt="Screenshot 2021-08-30 at 23 06 53" src="https://user-images.githubusercontent.com/402547/131369625-150a2055-6418-45f9-8eb2-358dc7ad771b.png">

It should be safe with old versions of the client, because they don't check for `data.error`, and if data.token is not set, they will just do nothing.

mifi commented:
I wonder why unit tests are failing on CI but not locally...

Murderlon commented:
You can merge this when you're ready :)

mifi commented:
Cool. Does the new string get translated before merge or will it happen after merge?

Murderlon commented:
> Cool. Does the new string get translated before merge or will it happen after merge? As in is there anything I need to do?

We only update English, the other locales are non-blocking for merge and mostly community contributions.

You can add this for Dutch: `Authenticatie geannuleerd`.

@aduh95 could suggest French.

mifi commented:
Gotcha. Are we using any translation service? For my app [LosslessCut](https://hosted.weblate.org/engage/losslesscut/) I use weblate, and it works quite well. Everyone is free to contribute through weblate's web UI which has translation memory and all, and weblate will automatically read english strings from github and periodically create an aggregated PR to update the other languages when people translate. It's free for open source.

aduh95 commented on packages/@uppy/companion/src/server/controllers/callback.js:
```diff
@@ -1,8 +1,24 @@
+  <head>
+      <meta charset="utf-8" />
+      <script>
+      window.opener.postMessage(${serialize({ error: true })}, ${serialize(origin)})
```
We may want to add a check here.
```suggestion
      window.opener && window.opener.postMessage(${serialize({ error: true })}, ${serialize(origin)})
```
Or we could leave it as is, and if it throws, the window doesn't close, the users sees the `Authentication failed.` in the body and that's maybe better. In that case a comment could be useful:
```suggestion
      ` +
      // if window.opener is nullish, we want the following line to throw to avoid
      // the window closing without informing the user.
      `window.opener.postMessage(${serialize({ error: true })}, ${serialize(origin)})
```

aduh95 commented on packages/@uppy/companion/src/server/controllers/callback.js:
```diff
@@ -1,8 +1,24 @@
+  <head>
+      <meta charset="utf-8" />
+      <script>
+      window.opener.postMessage(${serialize({ error: true })}, ${serialize(origin)})
```
nit: serializing `{ error: true }` is not useful here.
```suggestion
      window.opener.postMessage({ error: true }, ${serialize(origin)})
```

mifi replied:
I think it's nice to have serialize there. It could help signal to a future developer that it is dangerous to add code here. If there is no `serialize` and in the future someone adds more data here without first adding `serialize`, then they could introduce a XSS vulnerability.

aduh95 commented:
> @aduh95 could suggest French.

`Authentification interrompue`
